src/dmd/expression.d

```diff
@@ -5321,6 +5322,30 @@ extern (C++) final class DotVarExp : UnaExp
     override Expression toLvalue(Scope* sc, Expression e)
     {
         //printf("DotVarExp::toLvalue(%s)\n", toChars());
+        if (e1.op == TOK.this_ && sc.ctorflow.fieldinit.length && !(sc.ctorflow.callSuper & CSX.any_ctor))
+        {
+            if (VarDeclaration vd = var.isVarDeclaration())
+            {
+                auto ad = vd.isMember2();
+                if (ad && ad.fields.dim == sc.ctorflow.fieldinit.length)
+                {
+                    foreach (i, f; ad.fields)
+                    {
+                        if (f == vd)
+                        {
+                            if (!(sc.ctorflow.fieldinit[i] & CSX.this_ctor))
+                            {
+                                /* If the address of vd is taken, assume it is thereby initialized
```

Comment:
This needs to be documented somewhere, otherwise you may end up with surprising error messages. For example, this code:

immutable(int)* g;

```d
struct X { 
    int a = 10; 
    immutable this(int x)
    {   
            g = &a; 
            a = 42;    // error
    }   
}

void main()
{
    auto x = immutable X();
}
```

Leads to : `test.d(7): Error: immutable field a initialized multiple times` which is correct but not obvious

Reply:
I think we should point to the first initialization in the error message. And maybe a special message explaining why taking address is initialization.